### PR TITLE
pull-branch-strategy

### PR DESCRIPTION
**Pull request by aricahunter**
Added a pull-branch-strategy configuration parameter which is set using:
git town config --pull-strategy [merge | rebase]

which is used for when the branches are synced

implements #671

### CODE REVIEW

File: src/helpers/configuration.sh [hunk 1]
```diff
@@ -19,7 +19,11 @@ fi
 
 export MAIN_BRANCH_NAME=$(get_configuration main-branch-name)
 export PERENNIAL_BRANCH_NAMES=$(get_configuration perennial-branch-names)
-
+if is_pull_strategy_configured ; then
+  export PULL_BRANCH_STRATEGY=$(get_configuration pull-branch-strategy)
```
Comment (kevgo) [inline src/helpers/configuration.sh]: This makes two calls to `get_configuration`. Let's try to get it down to one.

File: src/helpers/configuration_helpers.sh [hunk 2]
```diff
@@ -110,8 +118,16 @@ function run_config_operation {
...
     elif [ "$operation" == "--reset" ]; then
       remove_all_configuration
+    elif [ "$operation" == "--pull-strategy" ]; then
+      if [ -z "$2" ] ; then
+        echo "usage: git town config --pull-strategy [ merge | rebase ]"
+      elif [ "$2" != 'merge' ] && [ "$2" != 'rebase' ] ; then
+        echo "The only acceptable pull strategy values are merge and rebase"
+      else 
+        store_configuration pull-branch-strategy "$2"
+      fi
```
Comment (charlierudolph) [inline src/helpers/configuration_helpers.sh]: Instead of adding logic here please deal with this similar to how the main branch is dealt with. See [src/git-town](https://github.com/Originate/git-town/blob/master/src/git-town#L47).

File: src/helpers/sync_helpers.sh [hunk 3]
```diff
@@ -18,11 +18,11 @@ function sync_branch_steps {
       echo "merge_tracking_branch"
       echo "merge $(parent_branch "$branch")"
     else
-      echo "rebase_tracking_branch"
+      echo "$PULL_BRANCH_STRATEGY""_tracking_branch"
```
Comment (charlierudolph) [inline src/helpers/sync_helpers.sh]: Quotes in the middle are unnecessary. Should be:
```
"$PULL_BRANCH_STRATEGY_tracking_branch"
```

File: src/helpers/sync_helpers.sh [hunk 4]
```diff
@@ -18,11 +18,11 @@ function sync_branch_steps {
...
 
       if [ "$branch" = "$MAIN_BRANCH_NAME" ] && [ "$(has_remote_upstream)" = true ]; then
         echo "fetch_upstream"
-        echo "rebase upstream/$MAIN_BRANCH_NAME"
+        echo "$PULL_BRANCH_STRATEGY upstream/$MAIN_BRANCH_NAME"
```
Comment (charlierudolph) [inline src/helpers/sync_helpers.sh]: I don't think this should be effected by `PULL_BRANCH_STRATEGY`. @kevgo thoughts?

**Comment by charlierudolph** [comment]
Code looks good but we still need feature tests for this. Both for configuring the rebase strategy and using it. I can help you setup some feature tests next week, when I get back from vacation. @kevgo may be able to help as well.

File: src/helpers/sync_helpers.sh (hunk 3 above)
Comment (aricahunter) [inline src/helpers/sync_helpers.sh]: I believe they are
in my terminal if I do:
export FOO="meow"
echo "$FOO_foo"
it prints nothing because FOO_foo is not a variable
however
echo "$FOO""_foo" 
gives the desired result of meow_foo

File: src/helpers/configuration_helpers.sh (hunk 2 above)
Comment (aricahunter) [inline src/helpers/configuration_helpers.sh]: The main branch is dealt with by calling functions that then have logic. Are you saying you want me to put the logic into a separate file?

File: src/helpers/sync_helpers.sh (hunk 3 above)
Comment (charlierudolph) [inline src/helpers/sync_helpers.sh]: Aww sorry forgot about that. I believe we want  "${FOO}_foo" then.

File: src/helpers/configuration_helpers.sh (hunk 2 above)
Comment (charlierudolph) [inline src/helpers/configuration_helpers.sh]: Not necessarily a separate file but a separate function yes.

File: src/helpers/sync_helpers.sh (hunk 4 above)
Comment (kevgo) [inline src/helpers/sync_helpers.sh]: This implements #671, which mentions that the main branch is also affected by this. I'm personally fine either way. I guess for consistency's sake it makes sense to merge the main branch as well when GT is configured to merge when pulling.

**Comment by kevgo** [comment]
@aricahunter this looks pretty good! Let's see how it looks like with the last 2 comments addressed!

File: src/helpers/sync_helpers.sh (hunk 3 above)
Comment (aricahunter) [inline src/helpers/sync_helpers.sh]: I think that tries to evaluate it as a function. I

Comment (charlierudolph) [inline src/helpers/sync_helpers.sh]: We have used that syntax before [here](https://github.com/Originate/git-town/blob/master/src/helpers/helpers.sh#L50-L51)

Comment (aricahunter) [inline src/helpers/sync_helpers.sh]: ah yah. I was using parens instead of curly braces.

**Comment by charlierudolph** [comment]
I believe the interface should be updated to 
```
git town pull-strategy [(merge | rebase)]
```
where 
* `git town pull-strategy` outputs the pull strategy
* `git town pull-strategy merge` sets the pull strategy to merge 
* `git town pull-strategy rebase` sets the pull strategy to rebase

**Comment by kevgo** [comment]
Hey @aricahunter! I have added some basic integration tests for your feature in https://github.com/Originate/git-town/commit/b5ed564647e1743d059263f2eb30d925c67940aa.

You can run them via `cucumber features/git-sync/current_branch/feature_branch/no_conflict/with_merge_pull_strategy.feature` and `cucumber features/git-town/configuration/pull_strategy/merge_option.feature`.

Hopefully this gives you enough momentum to add some more tests to cover more edge cases, like setting the merge strategy to another value, and how `git-sync` behaves in different situations with the "merge" pull strategy set. Feel free to ask questions if you get stuck!

File: src/helpers/sync_helpers.sh (hunk 4 above)
Comment (kevgo) [inline src/helpers/sync_helpers.sh]: Thinking about this some more, I'm now agreeing with @charlierudolph. The main branch should never be merged. (1) it only has a single commit on it, so there is no difference between merge and rebase, and (2) having merges on the main branch is ugly, and one of the goals of GT is to prevent exactly that.

**Comment by charlierudolph** [comment]
@kevgo I took this over and cleaned it up. Please review when you get a chance.

@aricahunter thanks for your help

File: features/git-town/configuration/pull_branch_strategy/display.feature [hunk 5]
```diff
@@ -0,0 +1,30 @@
+Feature: passing an invalid option to the pull strategy configuration
+
+  As a user or tool configuring Git Town's pull branch strategy
+  I want to know what the existing value for the pull-strategy is.
+
```
Comment (kevgo) [inline features/git-town/configuration/pull_branch_strategy/display.feature]: So that I can decide whether to adjust it.

File: features/git-town/configuration/pull_branch_strategy/display.feature [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  Scenario: default
```
Comment (kevgo) [inline features/git-town/configuration/pull_branch_strategy/display.feature]: Scenario: default setting

File: features/step_definitions/town_steps.rb [hunk 7]
```diff
@@ -61,6 +66,11 @@
 end
 
 
+Then(/^my repo is configured with the pull branch strategy "(.+?)"$/) do |pull_branch_strategy|
```
Comment (kevgo) [inline features/step_definitions/town_steps.rb]: Let's rename this step to "my repo is **now** configured with the pull branch strategy" to (a) imply that this step verifies existing state, and (b) distinguish it from "my repository has the "([^"]*)" pull branch strategy configured".

**Comment by kevgo** [comment]
Nice!! Thanks for doing this!  :ship:

**Comment by charlierudolph** [comment]
@kevgo. Addressed comments and then added in documentation that I nearly forgot. Please review the documentation updates (especially the README).

File: README.md [hunk 8]
```diff
@@ -57,6 +57,16 @@ Git Town is configured on a per-repository basis. It requires two pieces of info
 Upon first use in a repository, you will be prompted for this information.
 Use the [git town](/documentation/commands/git-town.md) command to view or update your configuration at any time.
 
+#### More Configuration
```
Comment (kevgo) [inline README.md]: "More" sounds a bit too generic. What about naming this "optional configuration"?

File: README.md [hunk 9]
```diff
@@ -57,6 +57,16 @@ Git Town is configured on a per-repository basis. It requires two pieces of info
...
+
+The following are also configured on a per-repository basis.
```
Comment (kevgo) [inline README.md]: The following configuration options have usable defaults, so the configuration wizard does not ask about them. 
Use the [git town](/documentation/commands/git-town.md) command to view or update them for a particular repository if needed.

File: documentation/commands/git-town.md [hunk 10]
```diff
@@ -51,5 +52,15 @@ git town version
 > git town perennial-branches --remove qa
 > ```
 
+* *pull-branch-strategy*
+> Displays the pull branch strategy. (default: 'rebase')
```
Comment (kevgo) [inline documentation/commands/git-town.md]: I would remove the period here (this is not a proper sentence), and add an empty like after this line.

File: documentation/commands/git-town.md [hunk 11]
```diff
@@ -51,5 +52,15 @@ git town version
...
+> This strategy is used when syncing a local feature branch with its upstream.
+>
+> Specify a strategy to set the pull branch strategy.
```
Comment (kevgo) [inline documentation/commands/git-town.md]: I would remove either this line or line 61. They are pretty much equivalent in their role to help understand what this example does.

File: man/man1/git-town.1 [hunk 12]
```diff
@@ -72,3 +74,9 @@ Add a new perennial branch
 
 .IP "git town perennial-branches --remove <branch_name>" 4
 Remove branch from perennial branches
+
+.IP "git town pull-branch-strategy" 4
+Displays your pull branch strategy
```
Comment (kevgo) [inline man/man1/git-town.1]: Displays the current pull branch strategy for the current repository

File: man/man1/git-town.1 [hunk 13]
```diff
@@ -72,3 +74,9 @@ Add a new perennial branch
...
+
+.IP "git town pull-branch-strategy <strategy>" 4
+Set your pull branch strategy as <strategy>
```
Comment (kevgo) [inline man/man1/git-town.1]: Set the given pull branch strategy for the current repository.
Allowed values are "merge" and "rebase".

**Comment by kevgo** [comment]
Ah, the documentation! I added some suggestions. Feel free to ship if you are happy with the outcome!

File: README.md (hunk 8 above)
Comment (charlierudolph) [inline README.md]: sounds good

File: README.md (hunk 9 above)
Comment (charlierudolph) [inline README.md]: The only change I would make is remove "usable". If a default isn't usable, it shouldn't be a default in my mind.

File: documentation/commands/git-town.md (hunk 11 above)
Comment (charlierudolph) [inline documentation/commands/git-town.md]: I'll remove line 61

File: man/man1/git-town.1 (hunk 12 above)
Comment (charlierudolph) [inline man/man1/git-town.1]: I don't think the first current is needed. Also none of the other commands reference the repository. The readme mentions git town is configured per repository. We could add something like that to this man page.

**Comment by kevgo** [comment]
Comments sound good, :ship:
